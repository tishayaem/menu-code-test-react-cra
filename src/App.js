import React from 'react';
import data from './menu-data.json';
import SelectDish from './SelectDish';
import Order from './Order';
export default class App extends React.Component {
  constructor() {
    super();
    this.state = {
      menu: {},
      currentOrder: {},
      order: {}
    };
  }

  componentDidMount() {
    this.setState({ menu: data });
  }
  handleRadio = event => {
    let course = event.target.name;
    let meal = this.state.menu[course].find(
      meal => meal.name === event.target.value
    );
    let currentOrder = this.state.currentOrder;
    currentOrder[course] = meal;

    this.setState({ currentOrder });
  };

  handleSubmit = event => {
    event.preventDefault();

    let order = this.state.currentOrder;
    this.setState({ order, currentOrder: {} });
  };

  render() {
    return (
      <div className="App">
        <SelectDish
          menu={this.state.menu}
          handleRadio={this.handleRadio}
          handleSubmit={this.handleSubmit}
        />
        <Order currentOrder={this.state.currentOrder} />
      </div>
    );
  }
}
