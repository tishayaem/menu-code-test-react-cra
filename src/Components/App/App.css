.interface {
  display: grid;
  grid-template-columns: repeat(3, 1fr);
  font-family: 'Roboto', sans-serif;
}

.menu {
  height: 100%;
}

.menu,
.current-order,
.final-order {
  font-family: 'Montserrat', sans-serif;
  padding: 30px;
}

.menu > form,
.current-order,
.final-order {
  display: flex;
  flex-direction: column;
  justify-content: space-between;
  height: 100%;
}

.error-message {
    width: 100%;
    background-color: rgb(184, 121, 121);
    font-size: 20px;
    padding: 10px;
    margin: 20px
}

