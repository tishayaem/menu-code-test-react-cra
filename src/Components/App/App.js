import React from 'react';
import data from '../../menu-data.json';
import SelectDish from '../SelectDish/SelectDish';
import Order from '../Order/Order';
import ErrorMessage from '../ErrorMessage/ErrorMessage';
import './App.css';
export default class App extends React.Component {
  constructor() {
    super();
    this.state = {
      menu: {},
      user: 1,
      currentOrder: {},
      order: {},
      error: false,
      areRadioChecked: null
    };
  }

  componentDidMount() {
    this.setState({ menu: data });
  }

  // Switch to uncontrolled buttons state
  changeRadio = () => {
    this.setState({ areRadioChecked: null });
  };

  handleRadio = event => {
    let course = event.target.name;
    let meal = this.state.menu[course].find(
      meal => meal.name === event.target.value
    );
    let currentOrder = this.state.currentOrder;
    currentOrder[course] = meal;
    this.setState({ currentOrder, error: false });
  };

  restaurantRules() {
    let { currentOrder, error, user, order } = this.state;

    let areTwoOrMoreOrdered = Object.keys(currentOrder).length < 2;
    let isMainOrdered = !currentOrder.mains;
   if(user === 2 && order[1].desserts && currentOrder.desserts) {
      let isCheesecakeavailable =  (order[1].desserts.name === 'Cheesecake' && currentOrder.desserts.name === 'Cheesecake')
        if (isCheesecakeavailable) {
          error = 'We are sorry, but cheesecakes are out of stock'
        }
    }

    else if (areTwoOrMoreOrdered) {
      error = 'Please order at least two dishes';
    } else if (isMainOrdered) {
      error = 'Please order main dish';
    } else if (currentOrder.starters && currentOrder.mains) {
      const doesPierreApprove =
        currentOrder.starters.name === 'Prawn cocktail' &&
        currentOrder.mains.name === 'Salmon fillet';

      if (doesPierreApprove) {
        error = "We are sorry, but you can't have Salmon fillet with prawn cocktail!";
      }
    }

    return error;
  }

  handleSubmit = event => {
    event.preventDefault();
    let { currentOrder, order, user } = this.state;
    let error = this.restaurantRules();
    if (error) {
      currentOrder = {};
    } else {
      order[user] = currentOrder;
      user += 1;
    }

    this.setState({
      order,
      currentOrder: {},
      error,
      user,
      areRadioChecked: false
    });
    // Temporary workaround to uncheck radio buttons on form submit. 
    // Must be replaced with better solution to take full control of radio buttons' state.
    if (user < 3) {setTimeout(() => { this.changeRadio()},0,1)}};

  render() {
    let {
      menu,
      areRadioChecked,
      user,
      order,
      currentOrder,
      error
    } = this.state;
    return (
      <div className="App">
      <div className="interface">
        <SelectDish
          menu={menu}
          handleRadio={this.handleRadio}
          handleSubmit={this.handleSubmit}
          areRadioChecked={areRadioChecked}
          user={user}
        />
        <Order currentOrder={currentOrder} order={order} user={user}/>
        <br/>
        </div>
        {error && (
         
          <div className= "error-message">
            <ErrorMessage error={error} />
          </div>
        )}
      </div>
    );
  }
}
