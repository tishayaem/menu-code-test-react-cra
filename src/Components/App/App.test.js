import React from 'react';
import App from './App';
import data from '../../menu-data.json';

describe('App', () => {
  const app = mount(<App />);

  it('renders properly', () => {
    expect(app).toMatchSnapshot();
  });

  it('renders menu from menu-data.json', () => {
    expect(app.find('.interface').exists()).toBe(true);
  });
});
