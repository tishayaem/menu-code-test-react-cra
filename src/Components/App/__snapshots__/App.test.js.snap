// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders properly 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <App />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="interface"
        >
          <SelectDish
            areRadioChecked={null}
            handleRadio={[Function]}
            handleSubmit={[Function]}
            menu={
              Object {
                "desserts": Array [
                  Object {
                    "id": 9,
                    "name": "Sticky toffee",
                    "price": 18,
                  },
                  Object {
                    "id": 10,
                    "name": "Tiramisu",
                    "price": 4.5,
                  },
                  Object {
                    "id": 11,
                    "name": "Cheesecake",
                    "price": 4,
                  },
                  Object {
                    "id": 12,
                    "name": "Ice cream",
                    "price": 3.5,
                  },
                ],
                "mains": Array [
                  Object {
                    "id": 5,
                    "name": "Steak",
                    "price": 18,
                  },
                  Object {
                    "id": 6,
                    "name": "Meatballs",
                    "price": 11.5,
                  },
                  Object {
                    "id": 7,
                    "name": "Salmon fillet",
                    "price": 14,
                  },
                  Object {
                    "id": 8,
                    "name": "Vegetarian lasagna",
                    "price": 12,
                  },
                ],
                "starters": Array [
                  Object {
                    "id": 1,
                    "name": "Soup",
                    "price": 3,
                  },
                  Object {
                    "id": 2,
                    "name": "Pâté",
                    "price": 5,
                  },
                  Object {
                    "id": 3,
                    "name": "Bruschetta",
                    "price": 4.5,
                  },
                  Object {
                    "id": 4,
                    "name": "Prawn cocktail",
                    "price": 6,
                  },
                ],
              }
            }
            user={1}
          />
          <Order
            currentOrder={Object {}}
            order={Object {}}
            user={1}
          />
          <br />
        </div>,
        false,
      ],
      "className": "App",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <SelectDish
              areRadioChecked={null}
              handleRadio={[Function]}
              handleSubmit={[Function]}
              menu={
                Object {
                  "desserts": Array [
                    Object {
                      "id": 9,
                      "name": "Sticky toffee",
                      "price": 18,
                    },
                    Object {
                      "id": 10,
                      "name": "Tiramisu",
                      "price": 4.5,
                    },
                    Object {
                      "id": 11,
                      "name": "Cheesecake",
                      "price": 4,
                    },
                    Object {
                      "id": 12,
                      "name": "Ice cream",
                      "price": 3.5,
                    },
                  ],
                  "mains": Array [
                    Object {
                      "id": 5,
                      "name": "Steak",
                      "price": 18,
                    },
                    Object {
                      "id": 6,
                      "name": "Meatballs",
                      "price": 11.5,
                    },
                    Object {
                      "id": 7,
                      "name": "Salmon fillet",
                      "price": 14,
                    },
                    Object {
                      "id": 8,
                      "name": "Vegetarian lasagna",
                      "price": 12,
                    },
                  ],
                  "starters": Array [
                    Object {
                      "id": 1,
                      "name": "Soup",
                      "price": 3,
                    },
                    Object {
                      "id": 2,
                      "name": "Pâté",
                      "price": 5,
                    },
                    Object {
                      "id": 3,
                      "name": "Bruschetta",
                      "price": 4.5,
                    },
                    Object {
                      "id": 4,
                      "name": "Prawn cocktail",
                      "price": 6,
                    },
                  ],
                }
              }
              user={1}
            />,
            <Order
              currentOrder={Object {}}
              order={Object {}}
              user={1}
            />,
            <br />,
          ],
          "className": "interface",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "areRadioChecked": null,
              "handleRadio": [Function],
              "handleSubmit": [Function],
              "menu": Object {
                "desserts": Array [
                  Object {
                    "id": 9,
                    "name": "Sticky toffee",
                    "price": 18,
                  },
                  Object {
                    "id": 10,
                    "name": "Tiramisu",
                    "price": 4.5,
                  },
                  Object {
                    "id": 11,
                    "name": "Cheesecake",
                    "price": 4,
                  },
                  Object {
                    "id": 12,
                    "name": "Ice cream",
                    "price": 3.5,
                  },
                ],
                "mains": Array [
                  Object {
                    "id": 5,
                    "name": "Steak",
                    "price": 18,
                  },
                  Object {
                    "id": 6,
                    "name": "Meatballs",
                    "price": 11.5,
                  },
                  Object {
                    "id": 7,
                    "name": "Salmon fillet",
                    "price": 14,
                  },
                  Object {
                    "id": 8,
                    "name": "Vegetarian lasagna",
                    "price": 12,
                  },
                ],
                "starters": Array [
                  Object {
                    "id": 1,
                    "name": "Soup",
                    "price": 3,
                  },
                  Object {
                    "id": 2,
                    "name": "Pâté",
                    "price": 5,
                  },
                  Object {
                    "id": 3,
                    "name": "Bruschetta",
                    "price": 4.5,
                  },
                  Object {
                    "id": 4,
                    "name": "Prawn cocktail",
                    "price": 6,
                  },
                ],
              },
              "user": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "currentOrder": Object {},
              "order": Object {},
              "user": 1,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {},
            "ref": null,
            "rendered": null,
            "type": "br",
          },
        ],
        "type": "div",
      },
      false,
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="interface"
          >
            <SelectDish
              areRadioChecked={null}
              handleRadio={[Function]}
              handleSubmit={[Function]}
              menu={
                Object {
                  "desserts": Array [
                    Object {
                      "id": 9,
                      "name": "Sticky toffee",
                      "price": 18,
                    },
                    Object {
                      "id": 10,
                      "name": "Tiramisu",
                      "price": 4.5,
                    },
                    Object {
                      "id": 11,
                      "name": "Cheesecake",
                      "price": 4,
                    },
                    Object {
                      "id": 12,
                      "name": "Ice cream",
                      "price": 3.5,
                    },
                  ],
                  "mains": Array [
                    Object {
                      "id": 5,
                      "name": "Steak",
                      "price": 18,
                    },
                    Object {
                      "id": 6,
                      "name": "Meatballs",
                      "price": 11.5,
                    },
                    Object {
                      "id": 7,
                      "name": "Salmon fillet",
                      "price": 14,
                    },
                    Object {
                      "id": 8,
                      "name": "Vegetarian lasagna",
                      "price": 12,
                    },
                  ],
                  "starters": Array [
                    Object {
                      "id": 1,
                      "name": "Soup",
                      "price": 3,
                    },
                    Object {
                      "id": 2,
                      "name": "Pâté",
                      "price": 5,
                    },
                    Object {
                      "id": 3,
                      "name": "Bruschetta",
                      "price": 4.5,
                    },
                    Object {
                      "id": 4,
                      "name": "Prawn cocktail",
                      "price": 6,
                    },
                  ],
                }
              }
              user={1}
            />
            <Order
              currentOrder={Object {}}
              order={Object {}}
              user={1}
            />
            <br />
          </div>,
          false,
        ],
        "className": "App",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <SelectDish
                areRadioChecked={null}
                handleRadio={[Function]}
                handleSubmit={[Function]}
                menu={
                  Object {
                    "desserts": Array [
                      Object {
                        "id": 9,
                        "name": "Sticky toffee",
                        "price": 18,
                      },
                      Object {
                        "id": 10,
                        "name": "Tiramisu",
                        "price": 4.5,
                      },
                      Object {
                        "id": 11,
                        "name": "Cheesecake",
                        "price": 4,
                      },
                      Object {
                        "id": 12,
                        "name": "Ice cream",
                        "price": 3.5,
                      },
                    ],
                    "mains": Array [
                      Object {
                        "id": 5,
                        "name": "Steak",
                        "price": 18,
                      },
                      Object {
                        "id": 6,
                        "name": "Meatballs",
                        "price": 11.5,
                      },
                      Object {
                        "id": 7,
                        "name": "Salmon fillet",
                        "price": 14,
                      },
                      Object {
                        "id": 8,
                        "name": "Vegetarian lasagna",
                        "price": 12,
                      },
                    ],
                    "starters": Array [
                      Object {
                        "id": 1,
                        "name": "Soup",
                        "price": 3,
                      },
                      Object {
                        "id": 2,
                        "name": "Pâté",
                        "price": 5,
                      },
                      Object {
                        "id": 3,
                        "name": "Bruschetta",
                        "price": 4.5,
                      },
                      Object {
                        "id": 4,
                        "name": "Prawn cocktail",
                        "price": 6,
                      },
                    ],
                  }
                }
                user={1}
              />,
              <Order
                currentOrder={Object {}}
                order={Object {}}
                user={1}
              />,
              <br />,
            ],
            "className": "interface",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "areRadioChecked": null,
                "handleRadio": [Function],
                "handleSubmit": [Function],
                "menu": Object {
                  "desserts": Array [
                    Object {
                      "id": 9,
                      "name": "Sticky toffee",
                      "price": 18,
                    },
                    Object {
                      "id": 10,
                      "name": "Tiramisu",
                      "price": 4.5,
                    },
                    Object {
                      "id": 11,
                      "name": "Cheesecake",
                      "price": 4,
                    },
                    Object {
                      "id": 12,
                      "name": "Ice cream",
                      "price": 3.5,
                    },
                  ],
                  "mains": Array [
                    Object {
                      "id": 5,
                      "name": "Steak",
                      "price": 18,
                    },
                    Object {
                      "id": 6,
                      "name": "Meatballs",
                      "price": 11.5,
                    },
                    Object {
                      "id": 7,
                      "name": "Salmon fillet",
                      "price": 14,
                    },
                    Object {
                      "id": 8,
                      "name": "Vegetarian lasagna",
                      "price": 12,
                    },
                  ],
                  "starters": Array [
                    Object {
                      "id": 1,
                      "name": "Soup",
                      "price": 3,
                    },
                    Object {
                      "id": 2,
                      "name": "Pâté",
                      "price": 5,
                    },
                    Object {
                      "id": 3,
                      "name": "Bruschetta",
                      "price": 4.5,
                    },
                    Object {
                      "id": 4,
                      "name": "Prawn cocktail",
                      "price": 6,
                    },
                  ],
                },
                "user": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "currentOrder": Object {},
                "order": Object {},
                "user": 1,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {},
              "ref": null,
              "rendered": null,
              "type": "br",
            },
          ],
          "type": "div",
        },
        false,
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
